// @dxc-technology/halstack-react/resultsetTable/ResultsetTable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof3 = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _variables = require("../common/variables.js");

var _Table = _interopRequireDefault(require("../table/Table"));

var _Paginator = _interopRequireDefault(require("../paginator/Paginator"));

var _useTheme = _interopRequireDefault(require("../useTheme"));

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof3(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function normalizeSortValue(sortValue) {
  return typeof sortValue === "string" ? sortValue.toUpperCase() : sortValue;
}

function sortArray(index, order, resultset) {
  return resultset.slice().sort(function (element1, element2) {
    var sortValueA = normalizeSortValue(element1[index].sortValue || element1[index].displayValue);
    var sortValueB = normalizeSortValue(element2[index].sortValue || element2[index].displayValue);
    var comparison = 0;

    if ((0, _typeof2["default"])(sortValueA) === "object") {
      comparison = -1;
    } else if ((0, _typeof2["default"])(sortValueB) === "object") {
      comparison = 1;
    } else if (sortValueA > sortValueB) {
      comparison = 1;
    } else if (sortValueA < sortValueB) {
      comparison = -1;
    }

    return order === "desc" ? comparison * -1 : comparison;
  });
}

var getMinItemsPerPageIndex = function getMinItemsPerPageIndex(currentPageInternal, itemsPerPage, page) {
  return currentPageInternal === 1 ? 0 : itemsPerPage * (page - 1);
};

var getMaxItemsPerPageIndex = function getMaxItemsPerPageIndex(minItemsPerPageIndex, itemsPerPage, resultset, page) {
  return minItemsPerPageIndex + itemsPerPage > resultset.length ? resultset.length : itemsPerPage * page - 1;
};

var ArrowUp = function ArrowUp() {
  return /*#__PURE__*/_react["default"].createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    height: "24",
    viewBox: "0 0 24 24",
    width: "24",
    fill: "currentColor"
  }, /*#__PURE__*/_react["default"].createElement("path", {
    d: "M0 0h24v24H0V0z",
    fill: "none"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    d: "M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
  }));
};

var ArrowDown = function ArrowDown() {
  return /*#__PURE__*/_react["default"].createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    height: "24",
    viewBox: "0 0 24 24",
    width: "24",
    fill: "currentColor"
  }, /*#__PURE__*/_react["default"].createElement("path", {
    d: "M0 0h24v24H0V0z",
    fill: "none"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    d: "M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
  }));
};

var BothArrows = function BothArrows() {
  return /*#__PURE__*/_react["default"].createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    height: "24",
    viewBox: "0 0 24 24",
    width: "24",
    fill: "currentColor"
  }, /*#__PURE__*/_react["default"].createElement("path", {
    d: "M0 0h24v24H0z",
    fill: "none"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    d: "M12 5.83L15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
  }));
};

var DxcResultsetTable = function DxcResultsetTable(_ref) {
  var columns = _ref.columns,
      rows = _ref.rows,
      _ref$showGoToPage = _ref.showGoToPage,
      showGoToPage = _ref$showGoToPage === void 0 ? true : _ref$showGoToPage,
      _ref$itemsPerPage = _ref.itemsPerPage,
      itemsPerPage = _ref$itemsPerPage === void 0 ? 5 : _ref$itemsPerPage,
      itemsPerPageOptions = _ref.itemsPerPageOptions,
      itemsPerPageFunction = _ref.itemsPerPageFunction,
      margin = _ref.margin,
      _ref$tabIndex = _ref.tabIndex,
      tabIndex = _ref$tabIndex === void 0 ? 0 : _ref$tabIndex;
  var colorsTheme = (0, _useTheme["default"])();

  var _useState = (0, _react.useState)(1),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      page = _useState2[0],
      changePage = _useState2[1];

  var _useState3 = (0, _react.useState)(""),
      _useState4 = (0, _slicedToArray2["default"])(_useState3, 2),
      sortColumnIndex = _useState4[0],
      changeSortColumnIndex = _useState4[1];

  var _useState5 = (0, _react.useState)("asc"),
      _useState6 = (0, _slicedToArray2["default"])(_useState5, 2),
      sortOrder = _useState6[0],
      changeSortOrder = _useState6[1];

  var minItemsPerPageIndex = (0, _react.useMemo)(function () {
    return getMinItemsPerPageIndex(page, itemsPerPage, page);
  }, [itemsPerPage, page]);
  var maxItemsPerPageIndex = (0, _react.useMemo)(function () {
    return getMaxItemsPerPageIndex(minItemsPerPageIndex, itemsPerPage, rows, page);
  }, [itemsPerPage, minItemsPerPageIndex, page, rows]);

  var goToPage = function goToPage(newPage) {
    changePage(newPage);
  };

  var changeSorting = function changeSorting(columnIndex) {
    changePage(1);
    changeSortColumnIndex(columnIndex);
    changeSortOrder(sortColumnIndex === "" || sortColumnIndex !== columnIndex ? "asc" : sortOrder === "asc" ? "desc" : "asc");
  };

  var getIconForSortableColumn = function getIconForSortableColumn(clickedColumnIndex) {
    return sortColumnIndex === clickedColumnIndex ? sortOrder === "asc" ? /*#__PURE__*/_react["default"].createElement(ArrowUp, null) : /*#__PURE__*/_react["default"].createElement(ArrowDown, null) : /*#__PURE__*/_react["default"].createElement(BothArrows, null);
  };

  (0, _react.useEffect)(function () {
    if (rows.length > 0) {
      changePage(1);
    } else {
      changePage(0);
    }
  }, [rows.length, itemsPerPage]);
  var sortedResultset = (0, _react.useMemo)(function () {
    return sortColumnIndex !== "" ? sortArray(sortColumnIndex, sortOrder, rows) : rows;
  }, [sortColumnIndex, sortOrder, rows]);
  var filteredResultset = (0, _react.useMemo)(function () {
    return sortedResultset && sortedResultset.slice(minItemsPerPageIndex, maxItemsPerPageIndex + 1);
  }, [sortedResultset, minItemsPerPageIndex, maxItemsPerPageIndex]);
  return /*#__PURE__*/_react["default"].createElement(_styledComponents.ThemeProvider, {
    theme: colorsTheme.table
  }, /*#__PURE__*/_react["default"].createElement(DxcResultsetTableContainer, {
    margin: margin
  }, /*#__PURE__*/_react["default"].createElement(TableContainer, null, /*#__PURE__*/_react["default"].createElement(_Table["default"], null, /*#__PURE__*/_react["default"].createElement(HeaderRow, null, /*#__PURE__*/_react["default"].createElement("tr", null, columns.map(function (column, index) {
    return /*#__PURE__*/_react["default"].createElement(TableHeader, {
      key: "tableHeader_".concat(index)
    }, /*#__PURE__*/_react["default"].createElement(HeaderContainer, {
      key: "headerContainer_".concat(index),
      onClick: function onClick() {
        return column.isSortable && changeSorting(index);
      },
      tabIndex: column.isSortable ? tabIndex : -1,
      isSortable: column.isSortable
    }, /*#__PURE__*/_react["default"].createElement(TitleDiv, {
      isSortable: column.isSortable
    }, column.displayValue), column.isSortable && /*#__PURE__*/_react["default"].createElement(SortIcon, null, getIconForSortableColumn(index))));
  }))), /*#__PURE__*/_react["default"].createElement(TableRowGroup, null, filteredResultset.map(function (cells, index) {
    return /*#__PURE__*/_react["default"].createElement("tr", {
      key: "resultSetTableCell_".concat(index)
    }, cells.map(function (cellContent, index) {
      return /*#__PURE__*/_react["default"].createElement("td", {
        key: "resultSetTableCellContent_".concat(index)
      }, cellContent.displayValue);
    }));
  })))), /*#__PURE__*/_react["default"].createElement(PaginatorContainer, null, /*#__PURE__*/_react["default"].createElement(_Paginator["default"], {
    totalItems: rows.length,
    itemsPerPage: itemsPerPage,
    itemsPerPageOptions: itemsPerPageOptions,
    itemsPerPageFunction: itemsPerPageFunction,
    currentPage: page,
    showGoToPage: showGoToPage,
    onPageChange: goToPage,
    tabIndex: tabIndex
  }))));
};

var TableContainer = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  & table {\n    table-layout: auto;\n  }\n"])));

var PaginatorContainer = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])([""])));

var TableRowGroup = _styledComponents["default"].tbody(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  > div:nth-child(1) {\n    position: absolute;\n    left: calc(50% - 68.5px);\n    bottom: calc(50% - 68.5px - 30px);\n  }\n  & tr {\n    height: ", ";\n  }\n"])), function (props) {
  return props.theme.rowHeight || "70px";
});

var SortIcon = _styledComponents["default"].div(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  top: 409px;\n  left: 390px;\n  height: 14px;\n  cursor: pointer;\n  color: ", ";\n\n  svg {\n    height: 100%;\n    width: 100%;\n  }\n"])), function (props) {
  return props.theme.sortIconColor;
});

var TitleDiv = _styledComponents["default"].div(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2["default"])(["\n  cursor: ", ";\n"])), function (props) {
  return props.isSortable && "pointer" || "default";
});

var TableHeader = _styledComponents["default"].th(_templateObject6 || (_templateObject6 = (0, _taggedTemplateLiteral2["default"])([""])));

var HeaderContainer = _styledComponents["default"].div(_templateObject7 || (_templateObject7 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  align-items: center;\n  justify-content: ", ";\n  width: fit-content;\n  :focus {\n    ", "\n  }\n"])), function (props) {
  return props.theme.headerTextAlign === "center" ? "center" : props.theme.headerTextAlign === "right" ? "flex-end" : "flex-start";
}, function (props) {
  return props.isSortable && "outline: #0095ff solid 2px; \n        outline-offset: 4px;";
});

var HeaderRow = _styledComponents["default"].thead(_templateObject8 || (_templateObject8 = (0, _taggedTemplateLiteral2["default"])(["\n  height: 60px;\n"])));

var DxcResultsetTableContainer = _styledComponents["default"].div(_templateObject9 || (_templateObject9 = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: ", ";\n  margin: ", ";\n  margin-top: ", ";\n  margin-right: ", ";\n  margin-bottom: ", ";\n  margin-left: ", ";\n"])), function (props) {
  return props.theme.fontSizeBase;
}, function (props) {
  return props.margin && (0, _typeof2["default"])(props.margin) !== "object" ? _variables.spaces[props.margin] : "0px";
}, function (props) {
  return props.margin && (0, _typeof2["default"])(props.margin) === "object" && props.margin.top ? _variables.spaces[props.margin.top] : "";
}, function (props) {
  return props.margin && (0, _typeof2["default"])(props.margin) === "object" && props.margin.right ? _variables.spaces[props.margin.right] : "";
}, function (props) {
  return props.margin && (0, _typeof2["default"])(props.margin) === "object" && props.margin.bottom ? _variables.spaces[props.margin.bottom] : "";
}, function (props) {
  return props.margin && (0, _typeof2["default"])(props.margin) === "object" && props.margin.left ? _variables.spaces[props.margin.left] : "";
});

var _default = DxcResultsetTable;
exports["default"] = _default;
