// @dxc-technology/halstack-react/text-input/Suggestions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _useTranslatedLabels = _interopRequireDefault(require("../useTranslatedLabels"));

var _BackgroundColorContext = _interopRequireDefault(require("../BackgroundColorContext"));

var _Suggestion = _interopRequireDefault(require("./Suggestion"));

var _Icons = _interopRequireDefault(require("./Icons"));

var _templateObject, _templateObject2, _templateObject3, _templateObject4;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var Suggestions = function Suggestions(_ref) {
  var id = _ref.id,
      value = _ref.value,
      suggestions = _ref.suggestions,
      visualFocusIndex = _ref.visualFocusIndex,
      highlightedSuggestions = _ref.highlightedSuggestions,
      searchHasErrors = _ref.searchHasErrors,
      isSearching = _ref.isSearching,
      suggestionOnClick = _ref.suggestionOnClick,
      getTextInputWidth = _ref.getTextInputWidth;
  var translatedLabels = (0, _useTranslatedLabels["default"])();
  var backgroundType = (0, _react.useContext)(_BackgroundColorContext["default"]);
  var listboxRef = (0, _react.useRef)(null);

  var _useState = (0, _react.useState)(null),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      styles = _useState2[0],
      setStyles = _useState2[1];

  (0, _react.useLayoutEffect)(function () {
    var _listboxRef$current, _visualFocusedOptionE;

    var visualFocusedOptionEl = listboxRef === null || listboxRef === void 0 ? void 0 : (_listboxRef$current = listboxRef.current) === null || _listboxRef$current === void 0 ? void 0 : _listboxRef$current.querySelectorAll("[role='option']")[visualFocusIndex];
    visualFocusedOptionEl === null || visualFocusedOptionEl === void 0 ? void 0 : (_visualFocusedOptionE = visualFocusedOptionEl.scrollIntoView) === null || _visualFocusedOptionE === void 0 ? void 0 : _visualFocusedOptionE.call(visualFocusedOptionEl, {
      block: "nearest",
      inline: "start"
    });
  }, [visualFocusIndex]);

  var handleResize = function handleResize() {
    setStyles({
      width: getTextInputWidth()
    });
  };

  (0, _react.useLayoutEffect)(function () {
    handleResize();
  }, [getTextInputWidth]);
  (0, _react.useEffect)(function () {
    window.addEventListener("resize", handleResize);
    return function () {
      window.removeEventListener("resize", handleResize);
    };
  }, [getTextInputWidth]);
  return /*#__PURE__*/_react["default"].createElement(SuggestionsContainer, {
    id: id,
    error: searchHasErrors ? true : false,
    onMouseDown: function onMouseDown(event) {
      event.preventDefault();
    },
    ref: listboxRef,
    role: "listbox",
    backgroundType: backgroundType,
    style: styles
  }, !isSearching && !searchHasErrors && suggestions.length > 0 && suggestions.map(function (suggestion, index) {
    return /*#__PURE__*/_react["default"].createElement(_Suggestion["default"], {
      key: "suggestion-".concat(index),
      id: "suggestion-".concat(index),
      value: value,
      onClick: suggestionOnClick,
      suggestion: suggestion,
      isLast: index === suggestions.length - 1,
      visuallyFocused: visualFocusIndex === index,
      highlighted: highlightedSuggestions
    });
  }), isSearching && /*#__PURE__*/_react["default"].createElement(SuggestionsSystemMessage, null, translatedLabels.textInput.searchingMessage), searchHasErrors && /*#__PURE__*/_react["default"].createElement(SuggestionsError, null, /*#__PURE__*/_react["default"].createElement(SuggestionsErrorIcon, {
    backgroundType: backgroundType
  }, _Icons["default"].error), translatedLabels.textInput.fetchingDataErrorMessage));
};

var SuggestionsContainer = _styledComponents["default"].ul(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  box-sizing: border-box;\n  max-height: 304px;\n  overflow-y: auto;\n  margin: 0;\n  padding: 0.25rem 0;\n  background-color: ", ";\n  border: 1px solid\n    ", ";\n\n  border-radius: 0.25rem;\n  box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1);\n  color: ", ";\n  font-family: ", ";\n  font-size: ", ";\n  font-style: ", ";\n  font-weight: ", ";\n"])), function (props) {
  return props.error ? props.theme.errorListDialogBackgroundColor : props.theme.listDialogBackgroundColor;
}, function (props) {
  return props.error ? props.backgroundType === "dark" ? props.theme.errorBorderColorOnDark : props.theme.errorListDialogBorderColor : props.theme.listDialogBorderColor;
}, function (props) {
  return props.theme.listOptionFontColor;
}, function (props) {
  return props.theme.fontFamily;
}, function (props) {
  return props.theme.listOptionFontSize;
}, function (props) {
  return props.theme.listOptionFontStyle;
}, function (props) {
  return props.theme.listOptionFontWeight;
});

var SuggestionsSystemMessage = _styledComponents["default"].span(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  padding: 0.25rem 1rem;\n  color: ", ";\n  line-height: 1.715em;\n"])), function (props) {
  return props.theme.systemMessageFontColor;
});

var SuggestionsErrorIcon = _styledComponents["default"].span(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-wrap: wrap;\n  align-content: center;\n  margin-right: 0.5rem;\n  height: 18px;\n  width: 18px;\n  color: ", ";\n"])), function (props) {
  return props.backgroundType === "dark" ? props.theme.errorIconColorOnDark : props.theme.errorIconColor;
});

var SuggestionsError = _styledComponents["default"].span(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  padding: 0.25rem 1rem;\n  align-items: center;\n  line-height: 1.715em;\n  color: ", ";\n"])), function (props) {
  return props.theme.errorListDialogFontColor;
});

var _default = /*#__PURE__*/_react["default"].memo(Suggestions);

exports["default"] = _default;
